Updating read.md

ByteBoarder opened the pull request:
Updating read.md as apart of updating overall K8s node tools.

ByteBoarder committed:
Update README.md

commiting changes to be compared to master in k8s node tools

ByteBoarder committed:
Update README.md

spell issue

ByteBoarder committed:
read.md and image files.

Updating read.md
updating image files.

ByteBoarder committed:
fixing ```

ByteBoarder committed:
further updates.

ByteBoarder committed:
removing us to be moregeneric.

removing us

wangzhen127 commented on README.md:
```diff
@@ -1,3 +1,137 @@
+### How to Taint Nodes in GKE
+
+In Google Kubernetes Engine (GKE), you can taint nodes in node pools using `kubectl` or the `gcloud` command-line tool.
+Note: it's recommended to taint through gcloud as kubectl isn't permanent. 
```
gcloud and kubectl tainting/untainting serve different purposes. Use gcloud to taint the node pool, and use kubectl to untaint individual nodes.
- gcloud tainting does a node-pool wide setting and always taint the node when the node is just created. This is helpful when the node pool is scaled up, or when the node pool is upgrade. This means, by default all the nodes are tainted after startup.
- kubectl untainting is targeting individual nodes after the Daemonset finishes the setup, so that scheduler can start putting pods on it. We should only use kubectl to untaint here actually, which represents a temporary runtime setting for this particular node after Daemonset finishes the configuration. And this temporary status is desired.

ByteBoarder replied:
Updated and fixed. PTAL when able.

wangzhen127 commented on README.md:
```diff
@@ -1,3 +1,273 @@
+
+# Node Configuration in GKE
+
+The below outlines a strategy for performing custom node configuration in Google Kubernetes Engine (GKE) using init containers. The goal is to apply node-level settings (like sysctl adjustments, software installations, or kernel parameter checks) *before* regular application workloads are scheduled onto those nodes. To achieve this isolation temporarily, we'll mark nodes as unschedulable.
```
we'll mark nodes as unshcedulable with taint until configuration is completed.

wangzhen127 commented on README.md:
```diff
@@ -1,3 +1,273 @@
+
+1.  **Taint:** Apply a **`taint`** to a node pool. This prevents regular pods (without matching **`tolerations`**) from being scheduled there, effectively reserving it.
+2.  **Configure:** Deploy a DaemonSet (like the one in Part 1) but add a **`tolerations`** block that specifically matches the taint applied. This ensures the configuration init container *only* runs on the isolated, tainted nodes.
+3.  **Untaint:** Remove the taint from the node pool, allowing any workload (subject to other scheduling rules) to be scheduled on the now-configured nodes.
```
A better way is removing the taint from individual nodes, not the node pool at once.

The source of truth of taints is at node pool level, meaning when the node pool scales up, or after upgrade, the new nodes always get the taint. This is the desired behavior, to prevent pod scheduling until config is done (and taint is removed).

So in the daemonset script, we can add a last step to remove taint of this individual node. Maybe add some jitter and some retry mechanism with exponential backoff will help. Because on a large node pool, when every daemonset pod on each node is trying to untaint itself, the API server may get overloaded.

ByteBoarder replied:
Good call out. re-configured.

ByteBoarder committed:
Update README.md

wangzhen127 commented on README.md:
````diff
@@ -1,3 +1,273 @@
+        ```
+
+3.  **Remove the Taint:**
+    * Once configuration is complete, remove the taint from the node pool to make it generally schedulable.
````
We actually want to always keep the taint on the node pool, just in case when the node pool scales up, new nodes always get the taint. Here we should use kubectl to remove taint from individual node. More easily, we can just do this in the daemonset after we verify the change has been made. Basically each node's daemonset pod untaint the node itself.

SergeyKanzhelev commented on README.md:
```diff
@@ -1,3 +1,273 @@
+
+# Node Configuration in GKE
+
+The below outlines a strategy for performing custom node configuration in Google Kubernetes Engine (GKE) using init containers. The goal is to apply node-level settings (like sysctl adjustments, software installations, or kernel parameter checks) *before* regular application workloads are scheduled onto those nodes. To achieve this isolation temporarily, we'll mark nodes as unshcedulable with tai ...
```
this is what will be implemented, not what we already have, correct?

SergeyKanzhelev replied:
I mean taint part

ByteBoarder replied:
This is for what we have today with our privileged DS approach for init containers. Do you have suggestions?

ByteBoarder committed:
Merge branch 'GoogleCloudPlatform:master' into bulk-ds-update

ByteBoarder committed:
Update README.md

ByteBoarder committed:
Merge branch 'bulk-ds-update' of https://github.com/ByteBoarder/read.md-update into bulk-ds-update

ByteBoarder committed:
Merge branch 'GoogleCloudPlatform:master' into master

ByteBoarder committed:
Resolve merge conflict by incorporating both suggestions

ByteBoarder committed:
Merge pull request #2 from ByteBoarder/bulk-ds-update

Bulk ds update

ByteBoarder committed:
Update README.md

fixing merge issues.

ByteBoarder committed:
Merge pull request #3 from ByteBoarder/bulk-ds-update

Update README.md

wangzhen127 commented:
/lgtm
/hold
Please wait for Sergey to lgtm as well since he left comments before.

ByteBoarder committed:
Bulk update.

Cleaning up the repo.
Removing older files
Updating with read.md
Updating containers.

ByteBoarder committed:
re-doing folder structure

Clearer folder structure and easier deploys by implementing raw.github implementation.
added .md where they were required.

ByteBoarder committed:
Merge pull request #4 from ByteBoarder/ds-update

Ds update

SergeyKanzhelev commented on README.md:
```diff
@@ -1,3 +1,330 @@
+    *   **Status monitoring:** Tools to check the health and status of the Kubelet service.
+    *   **Log analysis:**  Scripts to collect and analyze Kubelet logs for debugging.
+*   **Containerd:**  Tools related to containerd, the container runtime used by Kubelet to manage containers. This might include:
+    *   **Container image management:** Tools for pulling,cl inspecting, or managing container images on nodes.
```
```suggestion
    *   **Container image management:** Tools for pulling, inspecting, or managing container images on nodes.
```

SergeyKanzhelev commented on README.md:
```diff
@@ -1,3 +1,330 @@
+
+## Supporting Tools for Node Configuration
+
+A collection of tools, potentially packaged within a container image used by an init container, can help manage and configure various aspects of a Kubernetes node:
```
what potentially means in this context?

ByteBoarder replied:
In some sections of this repository, we have config maps, scripts and other various packages not packed within a container.

SergeyKanzhelev commented on README.md:
```diff
@@ -1,3 +1,330 @@
+            echo "Successfully set net.ipv4.ip_forward."
+          else
+            echo "Failed to set net.ipv4.ip_forward." >&2
+            exit 1 # Fail the init container if command fails
```
this will keep the Init container restarting forever. Maybe add a note about it in comment.

ByteBoarder replied:
Will add a note in the next update.

SergeyKanzhelev commented on README.md:
```diff
@@ -1,3 +1,330 @@
+        command: ["/bin/sh", "-c"]
+        args:
+        - |
+          echo "Attempting to set net.ipv4.ip_forward=1 on the host..."
```
it is a good practice to check if whatever we do needs to be done or was already configured

ByteBoarder replied:
Updated.
```suggestion
          echo "Attempting to set net.ipv4.ip_forward=1 on the host..."
```

SergeyKanzhelev commented on README.md:
```diff
@@ -1,3 +1,330 @@
+    *   **Log collection and analysis:**  Tools to gather logs from various node components and facilitate analysis. to encompass:
+    *   **Log collection and analysis:**  Tools to gather logs from various node components and facilitate analysis.
+
+### Part 1: Scheduling an Init Container for Node Modification
```
does this documentation intend to replace the https://cloud.google.com/kubernetes-engine/docs/tutorials/automatically-bootstrapping-gke-nodes-with-daemonsets?

cc: @shannonxtreme

ByteBoarder replied:
It does not intend to replace it, just have a github version of it.
However, we will work on re-factoring this entire page.

SergeyKanzhelev commented on README.md:
```diff
@@ -1,3 +1,330 @@
+        # Minimal container just to keep the Pod running after init succeeds
+        image: gcr.io/gke-release/pause:latest
+  updateStrategy:
+    type: RollingUpdate
```
is it required to be specified? Is there any significance of it, just for my understanding?

ByteBoarder replied:
Yes, we need a pause container. We can't just have a init container.

SergeyKanzhelev commented on README.md:
```diff
@@ -1,3 +1,330 @@
+This section outlines a more advanced and automated method for node configuration. The workflow uses a single, intelligent DaemonSet that not only applies the configuration to tainted nodes but also automatically removes the taint from the node once its job is complete. This approach is ideal for streamlining configuration changes across a node pool without manual intervention to make the nodes s ...
+
+The process leverages the kubectl binary available on GKE nodes and requires RBAC permissions for the DaemonSet's ServiceAccount to modify its own node object.
+#### Concept:
```
```suggestion

#### Concept:
```

SergeyKanzhelev commented on README.md:
````diff
@@ -1,3 +1,330 @@
+        ```bash
+        kubectl describe node <node-name-in-pool> | grep Taints
+        ```
+        You should see node.config/status=initializing:NoSchedule.
````
did you mean `stage=configuring`?

```suggestion
        You should see node.config/stage=configuring:NoSchedule.
```

SergeyKanzhelev commented on README.md:
````diff
@@ -1,3 +1,330 @@
+    Save the following content as `auto-untaint-daemonset.yaml`.
+
+```
+# WARNING: THIS MAKES YOUR NODES LESS SECURE. This DaemonSet runs as privileged.
````
not only priviledged, but also have the service account for untainting nodes, opening it up for the workload-steering attacks

ByteBoarder replied:
Agreed, it's my desire to not have an exhaustive list(yet) on all the ways this can cause issues.

SergeyKanzhelev commented on README.md:
```diff
@@ -1,3 +1,330 @@
+          path: /
+      initContainers:
+      - name: configure-and-untaint
+        image: ubuntu:22.04 # Using a standard container image.
```
examples we provide are best given with the gcr-hosted images, not ones from dockeerhub

ByteBoarder replied:
fair, changed.

SergeyKanzhelev commented on README.md:
```diff
@@ -1,3 +1,330 @@
+          # 2. UNTAINT THE NODE
+          # This command removes the taint from the node this pod is running on.
+          echo "--> Untainting node ${NODE_NAME} by removing taint ${TAINT_KEY}..."
+          if ! /host/home/kubernetes/bin/kubectl taint node "${NODE_NAME}" "${TAINT_KEY}:NoSchedule-"; then
```
this is interesting. I like the use of kubectl. How does kubectl knows to use the service account with the proper permissions? Mostly curious how it all hooked up

ByteBoarder replied:
SA binds to cluster role through the cluster role binding, When pod is executed, as the SA is bounded through the previous setup, is using the token SA inherited to execute the bin.

SergeyKanzhelev commented on README.md:
```diff
@@ -1,3 +1,330 @@
+      # The main container is minimal; it just keeps the pod running.
+      containers:
+      - name: pause-container
+        image: registry.k8s.io/pause:3.9
```
inconsistent from other daemonset. Other one used `gcr.io/gke-release/pause:latest`

ByteBoarder replied:
fair, changed.pause

SergeyKanzhelev commented on README.md:
Also run priviledged containers as Init containers when possible so the presence of high-priv process is time bound

ByteBoarder replied:
Good call out, listed below.

SergeyKanzhelev commented on containerd/container-insecure-registry/insecure-registry-config.yaml:
```diff
@@ -21,7 +21,7 @@ spec:
       hostPID: true
```
this file is referred from docs: https://cloud.google.com/kubernetes-engine/docs/how-to/access-private-registries-private-certificates#update-ds-both-models Rename will break the link

(also there is a link from some internal playbooks)

SergeyKanzhelev replied:
perhaps the change needs to be synchronized

cc @shannonxtreme

ByteBoarder replied:
I've worked for our own playbooks to re-work.

Once we have this go through, I'm working to update all the other links as well with @shannonxtreme and others

SergeyKanzhelev commented on containerd-http-proxy/README.md:
````diff
@@ -13,8 +13,13 @@ kubectl apply -f sample_configmap.yaml
 2. Deploy the daemonset in `configure_http_proxy.yaml`. As it has been specifically allowlisted for GKE Autopilot, this **manifest in this repo cannot be changed if you are deploying to GKE Autopilot mode clusters**.
 
 ```
-kubectl apply -f configure_http_proxy.yaml
````
this file is also linked from a few doc pages as far as I can tell

ByteBoarder replied:
Indeed, once we updated this markdown. I'll work to update the docs same day this is approved. Live change should be quick.

wangzhen127 commented:
/lgtm
